netengine/backends/snmp/airos.py

```diff
@@ -400,23 +397,75 @@ def RAM_free(self):
         free = self.get_value('1.3.6.1.4.1.10002.1.1.1.1.2.0')
         return int(free)
 
+    @property
+    def RAM_buffered(self):
+        """
+        Returns the buffered RAM of the device
+        """
+        buffered = self.get_value('1.3.6.1.4.1.10002.1.1.1.1.3.0')
+        return int(buffered)
+
+    @property
```

Comment:
Look at the output of the `free` command:

```
$ free
              total        used        free      shared  buff/cache   available
Mem:       16267996     7406636     2508788     1475504     6352572     7158444
Swap:       8388604       52600     8336004
```

So far I see we have total, free, then something weird, for buffered and cached we have two different values but `free` shows them as one thing.

Can you please double check and compare the output of these SNMP mibs with free?

Is there a way to know the availeble memory? See https://www.linuxatemyram.com/ for more info.

Reply:
@nemesisdesign is this output from an AirOS device? For me the output of `free` looks like this:
```
WA# free
             total         used         free       shared      buffers
Mem:         61960        37960        24000            0         2020
-/+ buffers:              35940        26020
Swap:            0            0            0
```

Here are the corresponding values via SNMP:
- RAM_total: `SNMPv2-SMI::enterprises.10002.1.1.1.1.1.0 = Gauge32: 61960`
- RAM_buffered: `SNMPv2-SMI::enterprises.10002.1.1.1.1.3.0 = Gauge32: 2020`
- RAM_free: `SNMPv2-SMI::enterprises.10002.1.1.1.1.2.0 = Gauge32: 24068`
- RAM_cached: `SNMPv2-SMI::enterprises.10002.1.1.1.1.4.0 = Gauge32: 0`

I'm a bit confused about the last one. On this old [thread](https://community.ui.com/questions/AP-Pro-become-disconnected-when-they-exhaust-RAM-on-3-2-1/497e0198-5727-4248-b1fa-8c13d6e41709) (and a few others in Ubiquiti forums) I found the following piece of info:
```
1.3.6.1.4.1.10002.1.1.1.1.1.0 # memTotal

1.3.6.1.4.1.10002.1.1.1.1.2.0 # memFree

1.3.6.1.4.1.10002.1.1.1.1.3.0 # memBuffer

1.3.6.1.4.1.10002.1.1.1.1.4.0 # memCached

1.3.6.1.4.1.10002.1.1.1.2.1.0 # swapTotal

1.3.6.1.4.1.10002.1.1.1.2.2.0 # swapFree
```

I haven't found anything related to available memory yet, found nothing in the standard MIBs.

PS: the output of `free` you shared looks similar to my OpenWRT device and in the openwrt backend you'll see exactly that data being collected.

Reply:
@purhan I see, then probably AirOS still uses an old linux kernel which doesn't support this, can you check what kernel version it uses?

What abotu OpenWRT? Do we have the information of the available memory on OpenWRT?

Reply:
@nemesisdesign AirOS I use for testing is running on version: `2.6.32.68`.
With OpenWRT also, I didn't find anything related to available ram even though the kernel supports it.
Here are some relevant OIDs I got from snmpwalk:
```
HOST-RESOURCES-MIB::hrStorageDescr.1 = STRING: Physical memory
HOST-RESOURCES-MIB::hrStorageDescr.3 = STRING: Virtual memory
HOST-RESOURCES-MIB::hrStorageDescr.6 = STRING: Memory buffers
HOST-RESOURCES-MIB::hrStorageDescr.7 = STRING: Cached memory
HOST-RESOURCES-MIB::hrStorageDescr.8 = STRING: Shared memory
HOST-RESOURCES-MIB::hrStorageDescr.10 = STRING: Swap space
```
I believe this means these are the only host resources we can use then?

Reply:
@purhan this may be the memory, did you compare these values to what you can get from `free` or `cat /proc/meminfo`?

Regarding AirOS I was not aware they used such an old kernel version. Did you already upgrade the AirOS firmware to the latest available version?

Reply:
> @purhan this may be the memory, did you compare these values to what you can get from `free` or `cat /proc/meminfo`?

Yes, the values match with the ones from `free`.
 
> Regarding AirOS I was not aware they used such an old kernel version. Did you already upgrade the AirOS firmware to the latest available version?

I have not upgraded yet, I will let you know more about this soon.

Reply:
I compared memory results from netengine and running free command on OpenWrt and found some discrepancy. 
![Screenshot from 2021-07-05 21-34-04](https://user-images.githubusercontent.com/32094433/124497864-24349680-ddd9-11eb-976d-ba99b06b9ae3.png)

While testing I ran the free command first for control, fetched resources using netengine and then executed the free command again in hope that I will get output from netengine somewhere between these two values. 

The value of free memory was very off in my testing. We need to make sure that results from SNMP is same as from free command. Maybe we'll need to do some calculation manually.

Reply:
@pandafy I guess that's just a drawback of SNMP? I did some research on this and found some [discussions](https://serverfault.com/questions/640459/snmp-memory-values-do-not-match-free) but no luck

Reply:
@pandafy @nemesisdesign I upgraded the device to the latest firmware (v8.7.4). Output from free command appears to be the same:
```
WA# free
             total         used         free       shared      buffers
Mem:         61960        45268        16692            0         4128
-/+ buffers:              41140        20820
Swap:            0            0            0
```
Kernel version is the same for this (i.e. 2.6.32.68)
I don't see a difference in the OIDs.
Tested netengine and the results are similar, so no noticeable difference yet.